src/library.js

```diff
@@ -2388,6 +2387,25 @@ mergeInto(LibraryManager.library, {
   _emscripten_get_now_is_monotonic__deps: ['$nowIsMonotonic'],
   _emscripten_get_now_is_monotonic: () => nowIsMonotonic,
 
+  __emscripten_atomics_sleep: (ms) => {
+    try {
+      Atomics.wait(waitBuffer, 0, 0, ms);
+      return 1;
+    } catch (_) {
+      return 0;
+    }
```

Comment:
Can we avoid relying on a try-catch? This is because when using a debugger, one may often need to break on caught exceptions, and this will then run into false positives and make the debugging flow more difficult.

Also there is a wait instruction in Wasm, so a JS function is not needed here.

Reply:
I think a problem of doing it in wasm is that the module will just not validate if threads are disabled. We want this to work in non-pthreads builds and even in VMs without threading support at all. JS at least allows graceful fallback using a try-catch.

You can set the debugger to ignore caught exceptions, which can help debugging, I think? Unless you have other ones you do care about...

Reply:
Ah right, good point on validation.



> You can set the debugger to ignore caught exceptions, which can help debugging, I think? Unless you have other ones you do care about...

The issue here is that Emscripten's [postamble try-catches all exceptions](https://github.com/emscripten-core/emscripten/blob/98cb77f225b0bfd19adf8a1b487987149a1f27c0/src/postamble.js#L87), so one always has to enable caught exceptions to be able to debug any throws inside the code.

In this instance it looks straightforward to replace the try-catch with an `if (ENVIRONMENT_IS_WORKER)` to avoid such issues.

Reply:
Do you think this is guaranteed to work in all Workers? I'm worried some older browsers might not support it.

We could perhaps check if `Atomics.wait` exists here instead of the try-catch?

Reply:
All browsers that implement `Atomics` will definitely also implement `.wait`. I never heard of a single implementation that would not have (briefly/intermittently) not had it.

Reply:
Checking for `ENVIRONMENT_IS_WORKER` is not correct because it will also work in node in the main thread, and potentially in other v8s that aren't the browser. I searched the spec and found no direct way to measure whether [`[[CanBlock]]`](https://262.ecma-international.org/10.0/#sec-agentcansuspend) is true or not. But we can use an `Atomics.wait` for 0 ms at startup and if that throws then we can set a boolean to say that we can't block. Then at runtime we won't need to throw.